src/core/plugin/luapi_application.h

```diff
@@ -327,6 +378,335 @@ static int applib_layerAction(lua_State* L) {
+        lua_getfield(L, 1, "tool");
+        lua_getfield(L, 1, "width");
+        lua_getfield(L, 1, "color");
+        lua_getfield(L, 1, "fill");
+        lua_getfield(L, 1, "lineStyle");
+
+        tool = luaL_optstring(L, -5, ""); // First thing, we're gonna need the tool type.
+
+        toolHandler = ctrl->getToolHandler();
+
+        // TODO: (willnilges) Handle DrawingType?
+        // TODO: (willnilges) Break out Eraser functionality into a new API call.
+
+        // Set tool type
+        if (strcmp("highlighter", tool) == 0) {
+            stroke->setToolType(STROKE_TOOL_HIGHLIGHTER);
+
+            size = toolSizeToString(toolHandler->getHighlighterSize());
+            thickness = toolHandler->getToolThickness(TOOL_HIGHLIGHTER)[toolSizeFromString(size)];
+
+            fillOpacity = toolHandler->getHighlighterFill();
+            filled = toolHandler->getHighlighterFillEnabled();
+
+            Tool& tool = toolHandler->getTool(TOOL_HIGHLIGHTER);
+            color = tool.getColor();
+            std::string drawingType = drawingTypeToString(tool.getDrawingType());
+        } else { // Tool type must be pen
+            if (!(strcmp("pen", tool) == 0))
+                g_warning("%s", FC(_F("Unknown stroke type: \"{1}\", defaulting to pen") % tool));
+
+            stroke->setToolType(STROKE_TOOL_PEN);
+
+            size         = toolSizeToString(toolHandler->getPenSize());
+            thickness    = toolHandler->getToolThickness(TOOL_PEN)[toolSizeFromString(size)];
+
+            fillOpacity  = toolHandler->getPenFill();
+            filled       = toolHandler->getPenFillEnabled();
+
+            Tool& tool   = toolHandler->getTool(TOOL_PEN);
+            color        = tool.getColor();
+            drawingType  = drawingTypeToString(tool.getDrawingType());
+            lineStyle    = StrokeStyle::formatStyle(tool.getLineStyle());
+        }
+
+        // Set width
+        if (lua_isnumber(L, -4)) // Check if the width was provided
+            stroke->setWidth(lua_tonumber(L, -4));
+        else
+            stroke->setWidth(thickness);
+
+        // Set color
+        if (lua_isinteger(L, -3)) // Check if the color was provided
+            stroke->setColor(Color(lua_tointeger(L, -3)));
+        else
+            stroke->setColor(color);
+
+        // Set fill
+        if (lua_isinteger(L, -2)) // Check if fill settings were provided
+            stroke->setFill(lua_tointeger(L, -2));
+        else if (filled)
+            stroke->setFill(fillOpacity);
+        else
+            stroke->setFill(-1); // No fill
+
+        // Set line style
+        if (lua_isstring(L, -1)) // Check if line style settings were provided
+            stroke->setLineStyle(StrokeStyle::parseStyle(lua_tostring(L, -1)));
+        else
+            stroke->setLineStyle(StrokeStyle::parseStyle(lineStyle.data()));
+
+        lua_pop(L, 5); // Finally done with all that Lua data.
+
+        // Add the Stroke
+        layer->addElement(stroke);
```

Comment:
I don't suppose there's a way to allow a whole bunch of strokes to be counted as one "undoing" action?

Reply:
What people want to do depends very much on the use case. If you add only a small number of strokes/splines, it may be more convenient, if all of them can be undone individually. In your use case of image transcription undoing the whole thing at once would be better.
I can't follow your link (is it valid?), but yes, `InsertsUndoAction` already supports undoing/redoing a bunch of elements at once. The only reasonable way of using that would be to pass all splines/strokes in a huge table 
```Lua
{
strokes = {stroke1, stroke2, ... }
allowUndoRedoAction = "together"      -- (or "individual" or "none")
}
```
(similarly with splines) where each stroke/spline in the table has the same shape as before, just without the `allowUndoRedoAction` key.
The functions would have to be renamed to `addStrokes` and `addSplines`, respectively.